src/circuits/gates.jl

```diff
@@ -455,3 +461,80 @@ gate(gn::String, s::Vector{<: Index}, ns::Int...; kwargs...) =
 ITensors.op(gn::GateName, ::SiteType"Qubit", s::Index...; kwargs...) =
   gate(gn, s...; kwargs...)
 
+
+
+"""
+    gate(M::Union{MPS,MPO}, gatename::String, site::Int; kwargs...)
+
+Generate a gate tensor for a single-qubit gate identified by `gatename`
+acting on site `site`, with indices identical to a reference state `M`.
+"""
+function gate(M::Union{MPS,MPO},
+              gatename::String,
+              site::Int; kwargs...)
+  site_ind = (typeof(M)==MPS ? siteind(M,site) :
+                               firstind(M[site], tags="Site", plev = 0))
+  return gate(gatename, site_ind; kwargs...)
+end
+
+"""
+    gate(M::Union{MPS,MPO},gatename::String, site::Tuple; kwargs...)
+
+Generate a gate tensor for a two-qubit gate identified by `gatename`
+acting on sites `(site[1],site[2])`, with indices identical to a 
+reference state `M` (`MPS` or `MPO`).
+"""
+function gate(M::Union{MPS,MPO},
+              gatename::String,
+              site::Tuple; kwargs...)
+  site_ind1 = (typeof(M)==MPS ? siteind(M,site[1]) :
+                                firstind(M[site[1]], tags="Site", plev = 0))
```

Comment:
This kind of code reminds me that it would be good to define a function `originalsiteind[s]` that returns the site index that would be used to create the MPS/MPO (and therefore is the "correct" site index to create new operators). Something like:
```julia
originalsiteind(M::MPS, n::Int) = siteind(M, n)
originalsiteind(M::MPO, n::Int) = dag(siteind(M, 1; plev = 0))
```
Notice the `dag` is important because for an MPO with QNs, the site index with prime level 0 is actually the daggered version of the site index used to create it. 

Not a priority for this PR, I can add an issue to implement this.

Reply:
Yes, could you please add the issue, thanks!